File: frontend/src/api/info.ts
import { request } from './request';

export type DualOffer = {
  price: number;
  offset: number;
  apr: number;
  duration: number;
};

export type DualOffersInfo = {
  currentTime: string;
  startTime: string;
  offers: DualOffer[];
};

export type Info = {
  btcPrice: number;
  dualOffersInfo?: DualOffersInfo;
};

export const requestGetInfo = () =>
  request<Info>({
    method: 'get',
    endpoint: '/info',
  });

export const requestUpdateDualOffersByHtml = (textBody: string) =>
  request<DualOffer[]>({
    method: 'post',
    endpoint: '/info/dualOffers/html',
    textBody,
  });

File: backend/src/db/models/info.py
from typing import List

from pydantic import BaseModel, ConfigDict


class DualOffer(BaseModel):
    price: float
    offset: float
    apr: float
    duration: int

    model_config = ConfigDict(exclude_unset=True)


class DualOffersInfo(BaseModel):
    currentTime: str
    startTime: str
    offers: List[DualOffer]

    model_config = ConfigDict(exclude_unset=True)


class Info(BaseModel):
    btcPrice: float
    dualOffersInfo: DualOffersInfo | None = None

    model_config = ConfigDict(exclude_unset=True)


def parseInfo(data: dict) -> Info:
    return Info.model_validate(data)
